### ISSUE TITLE
--read-special mode and known files cache

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#1241]
the known files cache must be avoided when in `--read-special` mode.

same size, mtime, inode are no indicator of "no change" for e.g. a block device.

**Comment by textshell** [comment]
I think the cache should only be avoided if the file to backup is actually a special file. Because i think it is plausible that users of borg use --read-special with a combination of directories to backup where only some contain block devices. i.e. a virtualisation host where the backup contains lvm volume groups and other data (log files, cd images, configuration, etc) in the same archive.

**Comment by enkore** [comment]
process_file():

```
is_regular_file = stat.S_ISREG(st.st_mode)
if is_regular_file:
    ids = cache.file_known_and_unchanged(path_hash, st, ignore_inode)
else:
    ids = None

...
if is_regular_file:
    cache.memorize_file(path_hash, st, [c[0] for c in chunks])
```

Possibly even...

```
if is_regular_file and if st.st_nlink > 1:
   ...
```

**Comment by textshell** [comment]
Yes, the first part is what i had in mind. I don’t have that code cached at the moment. st would be lstat of the path beeing backed up. What about symlinks, i guess it doesn’t matter if they are cached or not. But i might be wrong.

What is the idea about the st.st_nlink > 1?

**Comment by enkore** [comment]
in read special mode the symlinks are followed when stat'ing (#1216)

> What is the idea about the st.st_nlink > 1?

On second thought, wouldn't do aynthing, hardlinking device files and similar stuff should just work fine, they're just inodes after all.

**Comment by textshell** [comment]
#1241 is doing to be reverted or rewritten. TW is currently working on that. Because of #1214. But yes, st need to be what is backuped. So if a symlink is followed it would need to be a stat not a lstat. Maybe a fstat(fd) would be nicest.

**Comment by ThomasWaldmann** [comment]
https://github.com/borgbackup/borg/pull/1242 was merged into 1.0-maint.
